File: tsm-rf.c
/*
 * Rabinovich–Fabrikant System
 *
 * (c) 2018-2025 m4r35n357@gmail.com (Ian Smith), for licencing see the LICENCE file
 */
#include <stdio.h>
#include <stdlib.h>
#include "taylor-ode.h"

struct Parameters { real alpha, gamma; series _a, _b, _c; };

model *tsm_init_p (int argc, char **argv, int n) {
    CHECK(argc == 10);
    model *_ = malloc(sizeof (model)); CHECK(_);
    _->_a = tsm_jet(n);
    _->_b = tsm_jet(n);
    _->_c = tsm_jet(n);
    tsm_get_p(argv, argc, &_->alpha, &_->gamma);
    return _;
}

triplet ode (series x, series y, series z, const model *_, const int k) {
    _->_a[k] = z[k] + t_sqr(x, k) - t_const(1.0L, k);
    _->_b[k] = 4.0L * z[k] - _->_a[k];
    _->_c[k] = t_const(_->alpha, k) + t_mul(x, y, k);
    return (triplet) {
        .x = t_mul(y, _->_a, k) + _->gamma * x[k],
        .y = t_mul(x, _->_b, k) + _->gamma * y[k],
        .z = - 2.0L * t_mul(z, _->_c, k)
    };
}

File: tsm-rf.py
#!/usr/bin/env python3
#
#  (c) 2018-2023 m4r35n357@gmail.com (Ian Smith), for licencing see the LICENCE file

from sys import argv, stderr
from collections import namedtuple
from ad import Components, Context, tsm, tsm_jet, t_const, t_mul, t_sqr

class Parameters(namedtuple('ParametersType', ['α', 'γ', 'a', 'b', 'c'])):
    pass

def get_p(n):
    return Parameters(α=float(argv[8]), γ=float(argv[9]), a=tsm_jet(n), b=tsm_jet(n), c=tsm_jet(n))

def ode(x, y, z, p, k):
    p.a[k] = z[k] + t_sqr(x, k) - t_const(1.0, k)
    p.b[k] = 4.0 * z[k] - p.a[k]
    p.c[k] = t_const(p.α, k) + t_mul(x, y, k)
    return Components(x=t_mul(y, p.a, k) + p.γ * x[k],
                      y=t_mul(x, p.b, k) + p.γ * y[k],
                      z=- 2.0 * t_mul(z, p.c, k))


print(f'TSM: {argv}', file=stderr)
Context.places, order, h, steps = int(argv[1]), int(argv[2]), float(argv[3]), int(argv[4])  # controls
tsm(ode, Context.places, order, h, steps, float(argv[5]), float(argv[6]), float(argv[7]), get_p(order))
